File: Sources/FocusCore/FocusPageRule.swift
import Foundation

public struct FocusPageRule: Codable, Equatable, Sendable {
    public enum RunPhase: String, Codable, Sendable {
        case documentStart
        case documentEnd
    }

    public let id: String
    public let hosts: [String]
    public let pathPrefixes: [String]
    public let runPhase: RunPhase
    public let metaViewport: String?
    public let features: [FocusFeature]

    public init(
        id: String,
        hosts: [String] = [],
        pathPrefixes: [String] = [],
        runPhase: RunPhase,
        metaViewport: String? = nil,
        features: [FocusFeature]
    ) {
        self.id = id
        self.hosts = hosts
        self.pathPrefixes = pathPrefixes
        self.runPhase = runPhase
        self.metaViewport = metaViewport
        self.features = features
    }

    public func match(host: String, path: String) -> Bool {
        let matchesHost = hosts.isEmpty || hosts.contains(host)
        let normalizedPath = path.isEmpty ? "/" : path
        let matchesPath = pathPrefixes.isEmpty || pathPrefixes.contains { normalizedPath.hasPrefix($0) }
        return matchesHost && matchesPath
    }
}

File: focus-core/src/commonMain/kotlin/org/bilibilifocus/core/model/FocusPageRule.kt
package org.bilibilifocus.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class FocusPageRule(
    val id: String,
    val hosts: List<String> = emptyList(),
    val pathPrefixes: List<String> = emptyList(),
    val runPhase: RunPhase,
    val metaViewport: String? = null,
    val features: List<FocusFeature>,
) {
    @Serializable
    enum class RunPhase {
        @SerialName("documentStart")
        DOCUMENT_START,

        @SerialName("documentEnd")
        DOCUMENT_END;
    }

    fun match(host: String, path: String): Boolean {
        val matchesHost = hosts.isEmpty() || host in hosts
        val normalizedPath = if (path.isEmpty()) "/" else path
        val matchesPath = pathPrefixes.isEmpty() || pathPrefixes.any { normalizedPath.startsWith(it) }
        return matchesHost && matchesPath
    }
}
